PEC in enterprise

lutovich opened the pull request:
Move Property Existence Constraints to Neo4j Enterprise Edition.

lutovich committed:
Initial work on making PEC enterprise only

Commit:
Fixed maven project structure

Commit:
Added EnterpriseDatabase

lutovich committed:
More test fixes, added facilities for testing enterprise databases

nawroth committed:
Build configuration updates after moving Cypher docs

lutovich committed:
Fixed PEC tests in cypher-docs and refcard-tests modules

nawroth committed:
Fix for constraints documentation

thobe commented on community/kernel/src/main/java/org/neo4j/kernel/SchemaRuleVerifier.java:
```diff
@@ -17,26 +17,13 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
-package org.neo4j.kernel.impl.coreapi.schema;
+package org.neo4j.kernel;
```
This is most likely not the right package for this interface.

We do not want to include this interface in the public API, and types in this package become part of the public API.

How about making it a neighbour of `SchemaRule`, the thing that it verifies?

thobe committed:
Move PEC enforcement to enterprise

thobe committed:
Handle differing semantics in constraint handling

What really differs between editions is the semantics in how constraints
are handled. It therefore makes sense for the object that implements the
difference to talk about these semantics.
The things that differ are:

 * How we verify constraints.
 * What constraints we support reading from the store.
 * What constraints we support writing to the store.

The new ConstraintSemantics interface captures these concerns.

thobe committed:
Move PEC verification on creation to enterprise

thobe commented on community/kernel/src/main/java/org/neo4j/kernel/impl/constraints/StandardConstraintSemantics.java:
```diff
@@ -0,0 +1,97 @@
+    protected PropertyConstraint readNonStandardConstraint( PropertyConstraintRule rule )
+    {
+        // todo: message and new exception type
+        throw new IllegalStateException( "Property existence constraints can only be used on Neo4j enterprise" );
```
@akollegger could you have a look at the message here?

akollegger replied:
We should be more formal about the product names in written form, where here it should be "Neo4j Enterprise Edition". How about this short phrasing:

"Property existence constraint requires Neo4j Enterprise Edition"

For the exception type, this could either be a `ConstraintCreationFailure` or a new type like `CapabilityUnavailableException`.

thobe replied:
@akollegger this message is for when a database is loaded in Community Edition that has a property existence constraint in it (because it was created using Enterprise Edition), so `ConstraintCreationFailure` might not be an appropriate exception type here. Should the message still be the same?

thobe commented on community/kernel/src/main/java/org/neo4j/kernel/impl/constraints/StandardConstraintSemantics.java:
```diff
@@ -0,0 +1,97 @@
+    private IllegalStateException propertyExistenceConstraintsNotAllowed()
+    {
+        // todo: message and new exception type
+        return new IllegalStateException( "Property existence constraints can only be used on Neo4j enterprise" );
```
@akollegger could you have a look at the message here?

akollegger replied:
We should be more formal about the product names in written form, where here it should be "Neo4j Enterprise Edition". How about this short phrasing:

"Property existence constraint requires Neo4j Enterprise Edition"

For the exception type, this could either be a `ConstraintCreationFailure` or a new type like `CapabilityUnavailableException`.

thobe replied:
@akollegger this message is for when trying to create a constraint in Neo4j Community Edition.

thobe commented on community/kernel/src/main/java/org/neo4j/unsafe/batchinsert/BatchInserterImpl.java:
```diff
@@ -285,7 +284,8 @@ public Label apply( long from )
         indexStore = life.add( new IndexConfigStore( this.storeDir, fileSystem ) );
-        schemaCache = new SchemaCache( neoStore.getSchemaStore() );
+        // TODO: perhaps we should differentiate for BatchInserter running in enterprise here...
+        schemaCache = new SchemaCache( new StandardConstraintSemantics(), neoStore.getSchemaStore() );
```
Should we support creating PEC in BatchInserter?
If we should, then we need an EnterpriseModule here to get access to the EnterpriseConstraintSemantics, otherwise the "TODO" can be removed.

akollegger replied:
Let's defer creating PEC in BatchInserter, which is intended for high-speed creation of known-good data.

thobe commented:
Perhaps we should also clean up the modules, and introduce a new module in `enterprise` to contain PEC and Forseti, because those things really don't belong in `ha`... Any suggestions on a name for that module?

thobe commented:
The hipster build failed in the last CI run.

nawroth committed:
Constraint docs and moving modules

* Improvements for the constraint docs.
* Make the PDF build able to find the fonts needed.

thobe committed:
Provide nice error messages on constraint failure

Also add testing to ensure that Community Edition fails to start
when the database contains a PEC.

thobe committed:
Move PEC to an enterprise-kernel module

Since PEC is not related to HA, it makes sense for it to be in its
own module rather than in the HA module.

thobe committed:
Move Forseti to enterprise-kernel

thobe commented on enterprise/server-enterprise/src/main/java/org/neo4j/server/enterprise/EnterpriseNeoServer.java:
```diff
@@ -61,8 +72,16 @@ public EnterpriseNeoServer( Config config, Dependencies dependencies, LogProvide
+        {
+            return lifecycleManagingDatabase( ENTERPRISE_FACTORY );
+        }
+        return lifecycleManagingDatabase( COMMUNITY_FACTORY );
```
Should we really have the "community" database "mode" here?

thobe commented on enterprise/server-enterprise/src/main/java/org/neo4j/server/enterprise/EnterpriseNeoServer.java:
```diff
@@ -44,7 +45,8 @@
     public static final String SINGLE = "SINGLE";
     public static final String HA = "HA";
-    private static final GraphFactory ENTERPRISE_FACTORY = new GraphFactory()
+    public static final String ENTERPRISE = "ENTERPRISE";
```
I'm not convinced that we need or want to introduce a new server mode setting.

nawroth commented:
Attempting to fix the build, see #5231 
My changes are in https://github.com/nawroth/neo4j/commit/59cc9e3f80e9e0b23d777d44a0a2b76125cc09d4
Do they look sane?

systay commented:
Looks great, @nawroth. Let's close this one and work on #5231 instead

thobe commented:
I disagree, there are too many dependencies in that commit. I've updated https://github.com/lutovich/neo4j/tree/2.3-pec-in-enterprise, but cannot reopen this PR.

nawroth commented:
@thobe I was wondering but couldn't quickly see which dependencies were needed.

nawroth commented:
I created a new PR from the branch by @lutovich #5232
